src/Core/AppKernel.php

```diff
@@ -0,0 +1,416 @@
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Injector\InjectorLoader;
+use SilverStripe\Core\Injector\SilverStripeServiceConfigurationLocator;
+use SilverStripe\Core\Manifest\ClassLoader;
+use SilverStripe\Core\Manifest\ClassManifest;
+use SilverStripe\Core\Manifest\ModuleLoader;
+use SilverStripe\Core\Manifest\ModuleManifest;
+use SilverStripe\Dev\DebugView;
+use SilverStripe\Dev\Install\DatabaseAdapterRegistry;
+use SilverStripe\Logging\ErrorHandler;
+use SilverStripe\ORM\DB;
+use SilverStripe\View\ThemeManifest;
+use SilverStripe\View\ThemeResourceLoader;
+
+class AppKernel extends CoreKernel
+{
+    public function __construct()
+    {
+        // Initialise the dependency injector as soon as possible, as it is
+        // subsequently used by some of the following code
+        $injectorLoader = InjectorLoader::inst();
+        $injector = new Injector(array('locator' => SilverStripeServiceConfigurationLocator::class));
+        $injectorLoader->pushManifest($injector);
+        $this->setInjectorLoader($injectorLoader);
+
+        // Manifest cache factory
+        $manifestCacheFactory = $this->buildManifestCacheFactory();
+
+        // Class loader
+        $classLoader = ClassLoader::inst();
+        $classLoader->pushManifest(new ClassManifest(BASE_PATH, $manifestCacheFactory));
+        $this->setClassLoader($classLoader);
+
+        // Module loader
+        $moduleLoader = ModuleLoader::inst();
+        $moduleManifest = new ModuleManifest(BASE_PATH, $manifestCacheFactory);
+        $moduleLoader->pushManifest($moduleManifest);
+        $this->setModuleLoader($moduleLoader);
+
+        // Config loader
+        // @todo refactor CoreConfigFactory
+        $configFactory = new CoreConfigFactory($manifestCacheFactory);
+        $configManifest = $configFactory->createRoot();
+        $configLoader = ConfigLoader::inst();
+        $configLoader->pushManifest($configManifest);
+        $this->setConfigLoader($configLoader);
+
+        // Load template manifest
+        $themeResourceLoader = ThemeResourceLoader::inst();
+        $themeResourceLoader->addSet('$default', new ThemeManifest(
+            BASE_PATH,
+            project(),
+            $manifestCacheFactory
+        ));
+        $this->setThemeResourceLoader($themeResourceLoader);
+    }
+
+    public function getEnvironment()
+    {
+        // Check set
+        if ($this->enviroment) {
+            return $this->enviroment;
+        }
+
+        // Check saved session
+        $env = $this->sessionEnvironment();
+        if ($env) {
+            return $env;
+        }
+
+        // Check getenv
+        if ($env = getenv('SS_ENVIRONMENT_TYPE')) {
+            return $env;
+        }
+
+        return self::LIVE;
+    }
+
+    /**
+     * Check or update any temporary environment specified in the session.
+     *
+     * @return null|string
+     */
+    protected function sessionEnvironment()
+    {
+        // Check isDev in querystring
```

Comment:
Maybe use this and look at it in a separate PR, then? I think that the future is probably useful (if you need the post-response to run in dev mode, $_SESSION[isDev] is important), but it could be less hacky